Std/Data/Array/Lemmas.lean

```diff
@@ -844,3 +844,61 @@ instance [DecidableEq α] (a : α) (as : Array α) : Decidable (a ∈ as) :=
 /-! ### erase -/
 
 @[simp] proof_wanted erase_data [BEq α] {l : Array α} {a : α} : (l.erase a).data = l.data.erase a
+
+/-! ### swap -/
+
+@[simp] theorem get_swap_right (a : Array α) {i j : Fin a.size} :
+    (a.swap i j)[j.val]'(a.size_swap .. |>.symm ▸ j.isLt) = a[i] :=
```

Comment:
It seems unnatural to state this using `j.val` instead of just `j`.

@joehendrix or @digama0, do you know what the right approach is here? The author needs this version using `j.val` below, but the plain `j` version might also be useful. Is there a better alternative to duplicating the lemma?

Reply:
This statement now says:
```lean
@[simp] theorem get_swap_right (a : Array α) {i j : Fin a.size} : (a.swap i j)[j.val] = a[i] := sorry -- #mwe
```
(Removed `'(a.size_swap .. |>.symm ▸ j.isLt)` as @semorrison suggested)

Reply:
Part of the question is whether this theorem (and `get_swap_left`) should even be annotated with `@[simp]`. To be clear, the only reason I opted it use `j.val` instead of  `j` (as was originally intended) is because when I have `j` and tag this as a `@[simp]`, everything builds fine but I get the following Std4 linting error (When I remove the `@[simp]` tag this linting error also disappears)

```
-- Found 1 error in 3922 declarations (plus 16441 automatically generated ones) in Std with 13 linters

/- The `simpNF` linter reports:
SOME SIMP LEMMAS ARE NOT IN SIMP-NORMAL FORM.
see note [simp-normal form] for tips how to debug this.
https://leanprover-community.github.io/mathlib_docs/notes.html#simp-normal%20form -/
-- Std.Data.Array.Lemmas
./././Std/Data/Array/Lemmas.lean:850:9: error: Array.get_swap_right.{u_1} Left-hand side simplifies from
  (Array.swap a i j)[j]
to
  (Array.swap a i j)[↑j]
using
  simp only [@getElem_fin, @Array.data_length, @Fin.is_lt]
Try to change the left-hand side to the simplified term!
```

Reply:
It seems to me that the statement that has the `j.val` option is the most robust as is shown in the following #mwe:

```lean
import Std.Data.Array.Lemmas

@[simp] theorem Array.get_swap_right (a : Array α) {i j : Fin a.size} : (a.swap i j)[j.val] = a[i] := sorry -- if not already defined

example (a : Array α) (i j : Fin a.size) : sorry :=
  have _ : (a.swap i j)[j] = a[i] := by simp -- `j` version works
  have _ : (a.swap i j)[j] = a[i] := a.get_swap_right -- `j` version works

  sorry
```

Reply:
The comments both say `j` version works --- is that what you intended? That would give the opposite conclusion.

Reply:
The simpNF linter is usually right: it is telling you what the LHS should be, to avoid non-confluence.

Reply:
perhaps a comment can be added that says that this theorem can also be used to prove `|- (a.swap i j)[j] = a[i]`